R/kernel.r

```diff
@@ -99,11 +99,43 @@ handle_shell = function() {
         kernel_info(msg)
     } else if (msg$header$msg_type == "history_request") {
         history(msg)
+    } else if (msg$header$msg_type == "complete_request") {
+        complete(msg)
     } else {
         print(c("Got unhandled msg_type:", msg$header$msg_type))
     }
 },
 
+complete = function(request) {
+    # 5.0 protocol:
+    #code <- request$content$code
+    # once this is supported, we’ll need to change the
+    # .assignEnd call because it expects in-line pos
+    
+    # 4.0 protocol:
+    code <- request$content$line
+    
+    c.info <- with_top_env(executor$userenv, {
+        utils:::.assignLinebuffer(code)
+        utils:::.assignEnd(request$content$cursor_pos)
+        utils:::.guessTokenFromLine()
+        utils:::.completeToken()
+        
+        c(list(comps = utils:::.retrieveCompletions()),
+          utils:::.guessTokenFromLine(update = FALSE))
```

Comment:
I was a bit puzzled by what was going on here - is it worth a comment to explain that .guessTokenFromLine with update=FALSE returns the start and token fields? (it also surprised me that `update` changes the return type).